enhancements/catalog-data-transfer-reduction.md

```diff
@@ -128,295 +120,134 @@ Some terminology:
-        output.push(first)
-        output.push(last)
-      end
-    end
-  end
-
-end
-```
-
-#### Diff inputs and catalog snapshotting
-
-A DC makes sense as output‘s type because they represent all bundle metadata
-required to build an index image in one unified format,
-and `opm render` can consume any input format to produce
-a normalized declarative config which can be worked with internally.
-
-A declarative config makes less sense as `input_expand` since diff only
-requires catalog object identifiers to keep or remove their identifying
-objects from a catalog. More pertinent is the disconnected catalog's size:
-transferring an entire catalog from disconnected to connected is equally
-as un-scalable as connected to disconnected. Instead, a simple *catalog snapshot*,
-a list of packages containing channel and bundle elements, is easy
-to both produce from in-cluster subscriptions and manipulate by hand:
-
-```yaml
-packages:
-- name: foo
-  channels:
-    - name: preview
-      head: foo.v0.1.0-beta.2
-      versions:
-      - 0.1.0-beta.0
-      - 0.1.0-beta.1
-      - 0.1.0-beta.2
-    - name: stable
-      head: foo.v0.1.0
-      versions:
-      - 0.1.0
-- name: bar
-- name: baz
-  channels:
-...
-```
-
-Head information is only required if performing a FromHeads diff.
-
-Optionally a DC describing the full disconnected catalog state can be provided instead of a snapshot,
-and can be transformed into the snapshot format internally.
-
-A standalone tool or new `opm` command will be written to build a catalog snapshot
-from a cluster by making [`ListBundles()`][listbundles] gRPC calls to all registries running.
+Since package + bundle name is unique to the catalog,
+bundle lookup between two catalogs and comparison is possible.
+Two bundles are equal if all data and metadata of each bundle are semantically equivalent.
+If any datum semantically differs between the two, the bundles are unequal.
+Practically speaking, these data are bundle properties like channels,
+objects, and upgrade-related information.
+
+Given this primitive, two catalogs `A` and `A'` can be compared, where `A'` is considered
+a newer version of `A'`, and unequal bundles or bundles not in `A` can be added to some output
+catalog `B`. Then, every bundle `a'` in catalog `A'` either has a counterpart `a` in `A`
+that is either equal, or newer and therefore has a more state desirable,
+or has no counterpart and therefore is of the newest state:
+- If `a` exists and `a == a'`, neither `a` nor `a'` are added to `B` because `A` has newest bundle state `a`.
+- If `a` exists and `a != a'`, `a'` is added to `B` because `A` does not have the newest bundle state `a'`.
+- If `a` does not exist, `a'` is added to `B` because `A` does not have the newest bundle state `a'`.
+- If `a'` does not exist, `a` is not added to `B` because `A` has the newest bundle state `a`.
+
+The `headsOnly` mode is a special case of the above algorithm, because
```

Comment:
To supplement the user context here: the special case here is when a user mirrors for the first time: they just want the channel heads, not the entire history of all Operators ever released, since they are starting on a green field. 

Then, for subsequently triggered mirroring runs they would use `latest` to only download the difference between the initial mirroring run and whatever has been published in the meantime. Here we want to download the Operator history all the way back to the cannel heads of `A`, because we need that for update graphs to be complete.